e2e/metrics/exporter.go

```diff
@@ -0,0 +1,144 @@
+package metrics
+
+import (
+	"fmt"
+	"net/http"
+
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+)
+
+type ExporterTest struct {
+	url string
+}
+
+func NewExporterTest(url string) *ExporterTest {
+	return &ExporterTest{
+		url: url,
+	}
+}
+
+// getMetrics fetches the metrics from the Prometheus endpoint and returns them
+// as a map of metric families.
+func (et *ExporterTest) getMetrics() (map[string]*dto.MetricFamily, error) {
+	resp, err := http.Get(et.url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	var parser expfmt.TextParser
+	metrics, err := parser.TextToMetricFamilies(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	return metrics, nil
+}
+
+// FindOneMetric expects to find exactly one metric with the given name, resource name,
+// resource namespace, and labels. If no such metric is found, or if more than
+// one is found, an error is returned.
+//
+// `resourceName` and `resourceNamespace` are the values of the `name` and
+// `namespace` labels, respectively.
+//
+// If labels is nil, only the name and namespace labels are checked.
+func (m *ExporterTest) FindOneMetric(
+	metricName string,
+	labels map[string]string,
+) (*Metric, error) {
+	allMetrics, err := m.getMetrics()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get metrics: %w", err)
+	}
+
+	// Metric name exists.
+	mf, ok := allMetrics[metricName]
+	if !ok {
+		return nil, fmt.Errorf("metric %q not found", metricName)
+	}
+
+	var metrics []*dto.Metric
+	for _, metric := range mf.Metric {
+		m := Metric{Metric: metric}
+
+		// Check that all labels match, if present.
+		match := true
+		for k, v := range labels {
+			if m.LabelValue(k) != v {
+				match = false
+				break
+			}
+		}
+		if match {
+			metrics = append(metrics, metric)
+		}
+	}
+
+	if len(metrics) != 1 {
+		return nil, fmt.Errorf(
+			"expected to find 1 metric for %s{%s}, got %d",
+			metricName,
+			promLabels(labels),
+			len(metrics),
+		)
+	}
+
+	return &Metric{Metric: metrics[0]}, nil
+}
+
+type promLabels map[string]string
+
+func (l promLabels) String() string {
+	r := ""
+	for k, v := range l {
+		r += fmt.Sprintf("%s=%q, ", k, v)
```

Comment:
nit: not very important here since we're dealing with test code, but this could be done with a [strings.StringBuilder](https://pkg.go.dev/strings#Builder.WriteString).